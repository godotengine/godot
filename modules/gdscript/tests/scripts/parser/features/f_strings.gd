enum CustomEnum { ENUM_VALUE = 3 }

func test():
	print("non-formatted string {not a slot}")
	print(f"without slot")
	print(f"") # empty string
	print(f"with double quote ' delimiters")
	print(f'with single quote " delimiters')
	print(f"with bool literal {true} slot")
	print(f"with int literal {4} slot")
	print(f"with float literal {1.234} slot")
	print(f"with enum literal {CustomEnum.ENUM_VALUE} slot")
	print(f"with string literal {"foo"} slot")

	var bool_var: bool = true
	var int_var: int = 5
	var float_var: float = 5.234
	var enum_var: CustomEnum = CustomEnum.ENUM_VALUE
	var string_var: String = "foo"

	print(f"with bool variable {bool_var}")
	print(f"with int variable {int_var} slot")
	print(f"with float variable {float_var} slot")
	print(f"with enum variable {enum_var} slot")
	print(f"with string variable {string_var} slot")

	print(f"with bool expression {2 == 3+4}")
	print(f"with int expression {7*5-4} slot")
	print(f"with float expression {4.3*2 + 1.1} slot")
	print(f"with double quote string double quote expression {"foo"+"bar"} slot")
	print(f"with single quote string single quote expression {'foo'+'bar'} slot")
	print(f"with double quote string single quote expression {"foo"+'bar'} slot")
	print(f"with single quote string double quote expression {'foo'+"bar"} slot")

	print(f"with int function {double_me(5)} slot")
	print(f"with {1+2} multiple {1.1+2.2} slots")
	print(f"with slot { 1 +	2 } with whitespace inside")

	print(f"with {{escaped}} and not escaped {2+3} slots")
	print(fr"raw with {{escaped}} and not escaped {2+3} slots")
	print("non-formatted string doesn't honor {{escaped}} braces")
	print(f"with nested { f"formatted {2+3}" + " " + f"string {6*2}" } slots")

	print(fr"raw with raw \t chars and {2+3} slots")
	print(f"""multiline with newlines
and {2+3} slots""")

	# Any code tokens, including newlines, can go within a slot, even in
	# a non-multiline formatted string (probably not worth advertising this).
	print(f"with newlines and {
		"comments" # this is a comment
	} within a slot")

	print(fr'''raw and multiline with raw \t chars, carriage-returns
and {2+3} slots, using alternate quote char''')

	print(f"Handle slot with embedded braces { {"a":1, "b":2, "c":3}["b"]}")

	var s1 := "{1}"
	var s2 := "test"
	print(f"Handle slots that output String.format template targets {s1}:{s2}")

func double_me(x:int)->int:
	return x*2
