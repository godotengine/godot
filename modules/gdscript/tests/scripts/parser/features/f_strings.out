GDTEST_OK
non-formatted string {not a slot}
without slot

with double quote ' delimiters
with single quote " delimiters
with bool literal true slot
with int literal 4 slot
with float literal 1.234 slot
with enum literal 3 slot
with string literal foo slot
with bool variable true
with int variable 5 slot
with float variable 5.234 slot
with enum variable 3 slot
with string variable foo slot
with bool expression false
with int expression 31 slot
with float expression 9.7 slot
with double quote string double quote expression foobar slot
with single quote string single quote expression foobar slot
with double quote string single quote expression foobar slot
with single quote string double quote expression foobar slot
with int function 10 slot
with 3 multiple 3.3 slots
with slot 3 with whitespace inside
with {escaped} and not escaped 5 slots
raw with {escaped} and not escaped 5 slots
non-formatted string doesn't honor {{escaped}} braces
with nested formatted 5 string 12 slots
raw with raw \t chars and 5 slots
multiline with newlines
and 5 slots
with newlines and comments within a slot
raw and multiline with raw \t chars, carriage-returns
and 5 slots, using alternate quote char
Handle slot with embedded braces 2
Handle slots that output String.format template targets {1}:test
